crates/net/rlpx/connection.rs

```diff
@@ -18,180 +17,171 @@ pub const SUPPORTED_CAPABILITIES: [(&str, u8); 1] = [("p2p", 5)];
-    let frame_size = frame_data.len().to_be_bytes();
-    header.extend_from_slice(&frame_size[5..8]);
-
-    // header-data = [capability-id, context-id]  (both always zero)
-    let header_data = (0_u8, 0_u8);
-    header_data.encode(&mut header);
-
-    header.resize(16, 0);
-    egress_aes.apply_keystream(&mut header[..16]);
-
-    let header_mac_seed = {
-        let mac_digest: [u8; 16] = egress_mac.clone().finalize()[..16].try_into().unwrap();
-        let mut seed = mac_digest.into();
-        mac_aes_cipher.encrypt_block(&mut seed);
-        H128(seed.into()) ^ H128(header[..16].try_into().unwrap())
-    };
-    egress_mac.update(header_mac_seed);
-    let header_mac = egress_mac.clone().finalize();
-    header.extend_from_slice(&header_mac[..16]);
-
-    // Write header
-    stream.write_all(&header).await.unwrap();
-
-    // Pad to next multiple of 16
-    frame_data.resize(frame_data.len().next_multiple_of(16), 0);
-    egress_aes.apply_keystream(&mut frame_data);
-    let frame_ciphertext = frame_data;
-
-    // Send frame
-    stream.write_all(&frame_ciphertext).await.unwrap();
-
-    // Compute frame-mac
-    egress_mac.update(&frame_ciphertext);
-
-    // frame-mac-seed = aes(mac-secret, keccak256.digest(egress-mac)[:16]) ^ keccak256.digest(egress-mac)[:16]
-    let frame_mac_seed = {
-        let mac_digest: [u8; 16] = egress_mac.clone().finalize()[..16].try_into().unwrap();
-        let mut seed = mac_digest.into();
-        mac_aes_cipher.encrypt_block(&mut seed);
-        (H128(seed.into()) ^ H128(mac_digest)).0
-    };
-    egress_mac.update(frame_mac_seed);
-    let frame_mac = egress_mac.clone().finalize();
-
-    // Send frame-mac
-    stream.write_all(&frame_mac[..16]).await.unwrap();
-}
+    async fn write_frame(&mut self, mut frame_data: Vec<u8>) {
+        let egress_aes = &mut self.state.egress_aes;
+        let egress_mac = &mut self.state.egress_mac;
+        let mac_aes_cipher = Aes256Enc::new_from_slice(&self.state.mac_key.0).unwrap();
+
+        // header = frame-size || header-data || header-padding
+        let mut header = Vec::with_capacity(32);
+        let frame_size = frame_data.len().to_be_bytes();
+        header.extend_from_slice(&frame_size[5..8]);
+
+        // header-data = [capability-id, context-id]  (both always zero)
+        let header_data = (0_u8, 0_u8);
+        header_data.encode(&mut header);
+
+        header.resize(16, 0);
+        egress_aes.apply_keystream(&mut header[..16]);
+
+        let header_mac_seed = {
+            let mac_digest: [u8; 16] = egress_mac.clone().finalize()[..16].try_into().unwrap();
+            let mut seed = mac_digest.into();
+            mac_aes_cipher.encrypt_block(&mut seed);
+            H128(seed.into()) ^ H128(header[..16].try_into().unwrap())
+        };
+        egress_mac.update(header_mac_seed);
+        let header_mac = egress_mac.clone().finalize();
+        header.extend_from_slice(&header_mac[..16]);
+
+        // Write header
```

Comment:
I was thinking of what we talked earlier and I belive that a `Frame` struct is not very useful when reading a message, but it could be useful to encapsulate this `frame_header` and `frame_data` building to then write it to the stream in a single step with something like:
```Rust
self.stream.write_all(frame.encode())
```
WDYT? Of course this can also be done in a future PR.

Reply:
Yeah, I had the same idea but couldn't find a nice way to have this refactored... So far I only moved this code to a frame.rs module, but it is the same. I'll leave it this way to focus on the listen-loop... In any case this can be merged if it passes review and we can leave that refactor for a future PR